Client/WarFare/UILevelGuide.cpp

```diff
@@ -0,0 +1,385 @@
+	N3_VERIFY_UI_COMPONENT(m_pButtonCancel, (CN3UIButton*) GetChildByID("btn_cancel"));
+
+	return true;
+}
+
+void CUILevelGuide::LoadContent()
+{
+
+	int iSearchLevel;
+
+	//if user entered search level use it, else use user's current level
+	if (m_iSearchLevel == 0)
+		iSearchLevel = CGameBase::s_pPlayer->m_InfoBase.iLevel;
+	else
+		iSearchLevel = m_iSearchLevel;
+
+	//entered level cannot be bigger than max, cannot be smaller than min level
+	if (iSearchLevel < 1) iSearchLevel = 1;
+	if (iSearchLevel > m_iMaxLevel) iSearchLevel = m_iMaxLevel;
+
+	//update edit's text with searched level.
+	if (m_pEditLevel != nullptr)
+	{
+		std::string strLevel = std::to_string(iSearchLevel);
+		m_pEditLevel->SetString(strLevel.c_str());
+	}
+
+	//search by starting 5 levels lower and end 5 levels higher than user level.
+	int iLowerLevelLimit = iSearchLevel - m_iRangeSearch;
+	int iUpperLevelLimit = iSearchLevel + m_iRangeSearch;
+
+	if (iLowerLevelLimit <= 1) iLowerLevelLimit = 1;
+	if (iUpperLevelLimit <= 1) iUpperLevelLimit = 1;
+
+	if (iLowerLevelLimit >= m_iMaxLevel) iLowerLevelLimit = m_iMaxLevel;
+	if (iUpperLevelLimit >= m_iMaxLevel) iUpperLevelLimit = m_iMaxLevel;
+
+	int iCounter = 0;
+
+	size_t TableSize = CGameBase::s_pTbl_Quest_Content.GetSize();
+
+	std::fill(std::begin(m_saQuestTitle), std::end(m_saQuestTitle), "");
+	std::fill(std::begin(m_saQuestText), std::end(m_saQuestText), "");
+
+	for (size_t i = 0; i < TableSize && iCounter < m_iContentCount; i++)
+	{
+		__TABLE_QUEST_CONTENT* pQuestContent = CGameBase::s_pTbl_Quest_Content.GetIndexedData(i);
+
+		if (pQuestContent == nullptr)
+			continue;
+
+		int questLevel = pQuestContent->iLevel;
+
+		if (questLevel < iLowerLevelLimit || questLevel > iUpperLevelLimit)
+			continue;
+
+		m_saQuestTitle[iCounter] = pQuestContent->szTitle;
+		m_saQuestText[iCounter] = pQuestContent->szQuestText + "\n" + pQuestContent->szHintText;
+
+		iCounter++;
+	}
+
+	CN3UIString* pTitles[3] = { m_pTextTitle0, m_pTextTitle1, m_pTextTitle2 };
+	CN3UIString* pGuides[3] = { m_pTextGuide0, m_pTextGuide1, m_pTextGuide2 };
+
+	//find max number of pages, ceil without cmath
+	int iMaxPage = m_iContentCount / 3 + (m_iContentCount % 3 != 0 ? 1 : 0);
+
+	if (m_iCurrentPage < 1) m_iCurrentPage = 1;
+
+	if (m_iCurrentPage > iMaxPage) m_iCurrentPage = iMaxPage;
+
+	int iStartIndex = (m_iCurrentPage - 1) * 3;
+
+	for (int i = 0; i < 3; i++)
+	{
+
+		if (iStartIndex >= iCounter) //last page 2 empty items
+		{
+			pTitles[i]->SetString("");
+			pGuides[i]->SetString("");
+			continue;
+		}
+
+		if (pTitles[i])
+			pTitles[i]->SetString(m_saQuestTitle[iStartIndex]);
+
+		if (pGuides[i])
+			pGuides[i]->SetString(m_saQuestText[iStartIndex]);
+
+
+
+		iStartIndex++;
+	}
+
+	if (m_pTextPage != nullptr)
+	{
+		std::string strCurrentPage = std::to_string(m_iCurrentPage);
+		m_pTextPage->SetString(strCurrentPage);
+	}
+
+
+
+}
+
+bool CUILevelGuide::IsNumber(const std::string& s)
```

Comment:
I don't really like this being in `CUILevelGuide`. It's not really logic specific to this class whatsoever
Based on the existing usage this fits better in CN3UIEdit, buuuuuuut I really don't think it's altogether that necessary.

I think I'd simplify this by just going back to just atoi()'ing it -- you can keep it checking if the string is set, sure, but what would make more sense is filtering it on the control so it only allowed numerical input, rather than it allowing all input and hoping for the best.